### ISSUE TITLE
Link fails when using as a subproject (static library) (C++, MinGW, Windows)

### ISSUE THREAD

**Issue by starthal** [guijan/libobsd#2]
- Windows 10
 - GCC 11.2.0 installed via `scoop`
 - PowerShell

Minimal (?) reproduction:

subprojects/libobsd.wrap:
```ini
[wrap-git]
url = https://github.com/guijan/libobsd.git
revision = 9942023632f9c5a5b32db61cee6d37e996e0b1dd
depth = 1

[provide]
libbsd-overlay = libobsd_dep
libobsd = libobsd_dep
```

meson.build:
```meson
project('libobsd-test', ['cpp'])
libbsd_dep = dependency('libbsd-overlay')
test_src = files('main.cpp')
e = executable(
  'my-test',
  [test_src],
  dependencies : [libbsd_dep],
  install: false,
)    
test('my-test', e)
```

main.cpp:
```cpp
#include <string.h>

main(void)
{
	char buf[16];
	strlcat(buf, "another test string", sizeof(buf));
}
```

Output
```
PS C:\Users\i1B583116\src\libobsd-test> meson test -C .\builddir\ --suite 'libobsd-test'
ninja: Entering directory `C:\Users\i1B583116\src\libobsd-test\builddir'
[2/2] Linking target my-test.exe
FAILED: my-test.exe
"g++"  -o my-test.exe my-test.exe.p/main.cpp.obj "-Wl,--allow-shlib-undefined" "-Wl,--start-group" "subprojects\libobsd\libobsd.a" "-lbcrypt" "-lbcrypt" "-Wl,--subsystem,console" "-lkernel32" "-luser32" "-lgdi32" "-lwinspool" "-lshell32" "-lole32" "-loleaut32" "-luuid" "-lcomdlg32" "-ladvapi32" "-Wl,--end-group"
c:/users/_/scoop/apps/gcc/11.2.0/bin/../lib/gcc/x86_64-w64-mingw32/11.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: my-test.exe.p/main.cpp.obj:C:\Users\_\src\libobsd-test\builddir/../main.cpp:6: undefined reference to `__imp_strlcat'
collect2.exe: error: ld returned 1 exit status
```

The libobsd test builds also fail if `--suite` is not specified.

I believe this has something to do with the `-DOBSD_IMPORTING` flag and the resulting `dllimport` lines. When I prevent it from being added the build succeeds.

**Comment by guijan** [comment]
Thanks for the bug report. The issue is because `__declspec()` is for shared libraries only, but I made the mistake of using it for the static library too. I'm working on a solution, getting the compiler flags and `#defines` used inside my headers right across all the platforms is difficult.

**Comment by eli-schwartz** [comment]
It's a bit complicated, mostly due to Windows. Here's the basic approach I'd go with:

```h
/*
 * Here is the complicated part. Windows is special -- you cannot just define
 * entry points unconditionally.
 * */
#if defined(_WIN32) || defined(__CYGWIN__)
	/* When building static libraries, avoid marking public ones */
	#if defined(MYPROJECT_STATIC)
		#define MYPROJECT_EXPORT
	/* We are building this library */
	#elif defined(BUILDING_MYPROJECT)
		#define MYPROJECT_EXPORT __declspec(dllexport)
	/* We are using this library */
	#else
		#define MYPROJECT_EXPORT __declspec(dllimport)
	#endif

/* On to the easy part. GCC and lookalikes such as clang just work */
#elif defined(__GNUC__) && __GNUC__ >= 4
	#define MYPROJECT_EXPORT __attribute__((visibility("default")))

/* Older solaris support, why not */
#elif defined(__SUNPRO_C) && __SUNPRO_C >= 0x550
	#define MYPROJECT_EXPORT __global

/* All else failed, and we don't know about this compiler. Be conservative. */
#else
	#define MYPROJECT_EXPORT
#endif
```

On Windows, the MSVC convention rules supreme, and you must keep several states consistent:
- are you intending to make use of static linking? Don't use any declspec or you'll wreck everything.
- are you building the DLL? Mark exported functions.
- are you linking to the DLL? Mark those functions as imported.

Elsewhere, there's no confusion:
- are you using GCC or something else that supports the `__GNUC__` dialect? ***always*** just mark your functions as publicly interesting, let the compiler handle the rest
- Are you using older solaris compilers? They had a different way to say the exact same thing.

**Comment by guijan** [comment]
Fixed in 4364ee41fcadaf666127d323349557f32252f09b, there's also a test to prevent it from coming back again. Thank you again.
